components/properties/src/runtime.rs

```diff
@@ -0,0 +1,436 @@
+    /// String property `Simple_Titlecase_Mapping`
+    pub const SimpleTitlecaseMapping: Self = UnicodeProperty(0x4008);
+    /// String property `Simple_Uppercase_Mapping`
+    pub const SimpleUppercaseMapping: Self = UnicodeProperty(0x4009);
+    /// String property `Titlecase_Mapping`
+    pub const TitlecaseMapping: Self = UnicodeProperty(0x400A);
+    /// String property `Unicode_1_Name`
+    pub const Unicode1_Name: Self = UnicodeProperty(0x400B);
+    /// String property `Uppercase_Mapping`
+    pub const UppercaseMapping: Self = UnicodeProperty(0x400C);
+    /// String property `Bidi_Paired_Bracket`
+    pub const BidiPairedBracket: Self = UnicodeProperty(0x400D);
+
+    const STRING_MAX: Self = Self::BidiPairedBracket;
+
+    /// Misc property `Script_Extensions`
+    pub const ScriptExtensions: Self = UnicodeProperty(0x7000);
+}
+
+/// The kind of value a property has
+#[non_exhaustive]
+pub enum PropertyKind {
+    /// A binary property, has a value of True or False
+    Binary,
+    /// An enumerated (integer) property, has an integer value that is one amongst
+    /// a limited set of possible values for that property
+    Enumerated,
+    /// A mask property that is intended to be used as a bitmask
+    ///
+    /// Currently only [`UnicodeProperty::GeneralCategoryMask`] is in this group
+    Mask,
+    /// A double property, with a property that is a double-precision integer
+    ///
+    /// Currently only [`UnicodeProperty::NumericValue`] is in this group
+    Double,
+    /// A string property
+    String,
+    /// Miscellaneous property `ScriptExtensions`
+    ScriptExtensions,
+}
+
+impl UnicodeProperty {
+    /// Obtain the [`PropertyKind`] of this property.
+    ///
+    /// Returns `None` for unknown properties
+    pub fn kind(&self) -> Option<PropertyKind> {
+        let kind = if *self >= Self::Alphabetic && *self <= Self::BINARY_MAX {
+            PropertyKind::Binary
+        } else if *self <= Self::BidiClass && *self >= Self::ENUMERATED_MAX {
+            PropertyKind::Enumerated
+        } else if *self == Self::GeneralCategoryMask {
+            PropertyKind::Mask
+        } else if *self == Self::NumericValue {
+            PropertyKind::Double
+        } else if *self >= Self::Age && *self <= Self::STRING_MAX {
+            PropertyKind::String
+        } else if *self == Self::ScriptExtensions {
+            PropertyKind::ScriptExtensions
+        } else {
+            return None;
+        };
+        Some(kind)
+    }
+
+    /// Given a property name (long, short, or alias), returns the corresponding [`UnicodeProperty`]
+    /// value for it provided it belongs to the [subset relevant for ECMA402 regexes][subset]
+    ///
+    /// Returns none if the name does not match any of the names in this subset. Performs
+    /// strict matching of names.
+    ///
+    /// If using this to implement an ECMA402-compliant regex engine, please note these caveats:
```

Comment:
I think I have struck the right chord here with having an ECMA402-specific API that still makes sense in a general unicode context (especially; almost all the stuff I've listed is handled directly in ECMA402 spec text). But worth calling out